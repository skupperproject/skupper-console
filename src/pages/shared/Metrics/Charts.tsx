import { FC, memo } from 'react';

import { ChartThemeColor } from '@patternfly/react-charts';
import { Card, CardBody, CardTitle, GridItem, Icon } from '@patternfly/react-core';
import { CircleIcon } from '@patternfly/react-icons';
import { TableComposable, Tbody, Td, Th, Thead, Tr } from '@patternfly/react-table';

import { AvailableProtocols } from '@API/REST.enum';
import EmptyData from '@core/components/EmptyData';
import SkChartArea from '@core/components/SkChartArea';
import SkChartPie from '@core/components/SkChartPie';
import SkCounterCard from '@core/components/SkCounterCard';
import { convertToPercentage } from '@core/utils/convertToPercentage';
import { formatByteRate, formatBytes } from '@core/utils/formatBytes';
import { formatLatency } from '@core/utils/formatLatency';
import { formatToDecimalPlacesIfCents } from '@core/utils/formatToDecimalPlacesIfCents';
import { ProcessesLabels } from '@pages/Processes/Processes.enum';

import { MetricsLabels } from './Metrics.enum';
import { Metrics } from './services/services.interfaces';

const Charts: FC<{ metrics: Metrics; protocol?: AvailableProtocols }> = memo(({ metrics, protocol }) => (
  <>
    <GridItem span={8} rowSpan={2}>
      <Card isFullHeight>
        {!metrics.trafficDataSeriesPerSecond && <EmptyData />}
        {!!metrics.trafficDataSeriesPerSecond && (
          <>
            <CardTitle>{ProcessesLabels.ChartProcessDataTrafficSeriesAxisYLabel}</CardTitle>
            <CardBody>
              <SkChartArea
                themeColor={ChartThemeColor.orange}
                formatY={formatByteRate}
                legendLabels={[MetricsLabels.TrafficReceived, MetricsLabels.TrafficSent]}
                data={[
                  metrics.trafficDataSeriesPerSecond.timeSeriesDataReceived,
                  metrics.trafficDataSeriesPerSecond.timeSeriesDataSent
                ]}
              />
            </CardBody>
          </>
        )}
      </Card>
    </GridItem>

    {/* Total Traffic card */}
    <GridItem span={4}>
      {!metrics.trafficDataSeriesPerSecond && <EmptyData />}
      {!!metrics.trafficDataSeriesPerSecond && (
        <Card>
          <TableComposable borders={false} variant="compact">
            <Thead noWrap>
              <Tr>
                <Th />
                <Th>{MetricsLabels.ByteRateMaxCol}</Th>
                <Th>{MetricsLabels.ByteRateAvgCol}</Th>
                <Th>{MetricsLabels.ByteRateCurrentCol}</Th>
              </Tr>
            </Thead>
            <Tbody>
              <Tr>
                <Td>
                  <Icon size="sm">
                    <CircleIcon color={`var(--pf-chart-theme--orange--ColorScale--100, #ec7a08)`} />
                  </Icon>{' '}
                  {MetricsLabels.TrafficReceived}
                </Td>
                <Th>{formatByteRate(metrics.trafficDataSeriesPerSecond.maxTrafficReceived)}</Th>
                <Th>{formatByteRate(metrics.trafficDataSeriesPerSecond.avgTrafficReceived)}</Th>
                <Th>{formatByteRate(metrics.trafficDataSeriesPerSecond.currentTrafficReceived)}</Th>
              </Tr>
              <Tr>
                <Td>
                  <Icon size="sm">
                    <CircleIcon color={`var(--pf-chart-theme--orange--ColorScale--200, #f4b678)`} />
                  </Icon>{' '}
                  {MetricsLabels.TrafficSent}
                </Td>
                <Th>{formatByteRate(metrics.trafficDataSeriesPerSecond.maxTrafficSent)}</Th>
                <Th>{formatByteRate(metrics.trafficDataSeriesPerSecond.avgTrafficSent)}</Th>
                <Th>{formatByteRate(metrics.trafficDataSeriesPerSecond.currentTrafficSent)}</Th>
              </Tr>
            </Tbody>
          </TableComposable>
        </Card>
      )}
    </GridItem>

    {/* Chart pie for the distribution of the data traffic */}
    <GridItem span={4}>
      <Card isFullHeight>
        <SkChartPie
          format={formatBytes}
          themeColor={ChartThemeColor.orange}
          data={[
            {
              x: MetricsLabels.TrafficReceived,
              y: metrics.trafficDataSeries?.totalDataReceived || 0
            },
            {
              x: MetricsLabels.TrafficSent,
              y: metrics.trafficDataSeries?.totalDataSent || 0
            }
          ]}
        />
      </Card>
    </GridItem>

    {/* Chart latencies time series card*/}
    {protocol !== AvailableProtocols.Tcp && metrics.latencies && (
      <>
        <GridItem span={12}>
          <Card isFullHeight>
            <>
              <CardTitle>{ProcessesLabels.ChartProcessLatencySeriesAxisYLabel}</CardTitle>
              <CardBody>
                <SkChartArea
                  formatY={formatLatency}
                  themeColor={ChartThemeColor.multi}
                  legendLabels={metrics.latencies.timeSeriesLatencies.map(({ label }) => label)}
                  data={metrics.latencies.timeSeriesLatencies.map(({ data }) => data)}
                />
              </CardBody>
            </>
          </Card>
        </GridItem>

        {/* Chart requests time series card*/}
        {!!metrics.requestPerSecondSeries && (
          <>
            <GridItem span={8} rowSpan={2}>
              <Card isFullHeight>
                <>
                  <CardTitle>{MetricsLabels.RequestsPerSecondsSeriesAxisYLabel}</CardTitle>
                  <CardBody>
                    <SkChartArea
                      formatY={(y: number) => formatToDecimalPlacesIfCents(y, 3)}
                      themeColor={ChartThemeColor.purple}
                      legendLabels={metrics.requestPerSecondSeries.map(({ label }) => label)}
                      data={metrics.requestPerSecondSeries.map(({ data }) => data)}
                    />
                  </CardBody>
                </>
              </Card>
            </GridItem>

            {/*  Partial total request card*/}
            {!!metrics.requestSeries && (
              <GridItem span={4}>
                <SkCounterCard
                  title={'Total Requests'}
                  value={metrics.requestSeries[0].totalRequestInterval}
                  bgColor={'--pf-global--palette--purple-400'}
                  colorChart={ChartThemeColor.purple}
                  showChart={false}
                />
              </GridItem>
            )}

            {/*  avg request per second card*/}
            {!!metrics.requestPerSecondSeries && (
              <GridItem span={4}>
                <SkCounterCard
                  title={'Avg. Request rate'}
                  value={metrics.requestPerSecondSeries[0].avgRequestRateInterval}
                  bgColor={'--pf-global--palette--purple-200'}
                  showChart={false}
                />
              </GridItem>
            )}
          </>
        )}

        {/*  response card*/}
        {!!metrics.responseSeries && (
          <>
            <GridItem span={3}>
              <SkCounterCard
                title={metrics.responseSeries.statusCode2xx.label}
                value={
                  convertToPercentage(metrics.responseSeries.statusCode2xx.total, metrics.responseSeries.total) || ' - '
                }
                bgColor={'--pf-global--palette--green-400'}
                showChart={false}
              />
            </GridItem>
            <GridItem span={3}>
              <SkCounterCard
                title={metrics.responseSeries.statusCode3xx.label}
                value={
                  convertToPercentage(metrics.responseSeries.statusCode3xx.total, metrics.responseSeries.total) || ' - '
                }
                bgColor={'--pf-global--palette--blue-400'}
                showChart={false}
              />
            </GridItem>
            <GridItem span={3}>
              <SkCounterCard
                title={metrics.responseSeries.statusCode4xx.label}
                value={
                  convertToPercentage(metrics.responseSeries.statusCode4xx.total, metrics.responseSeries.total) || ' - '
                }
                bgColor={'--pf-global--palette--orange-100'}
                showChart={false}
              />
            </GridItem>
            <GridItem span={3}>
              <SkCounterCard
                title={metrics.responseSeries.statusCode5xx.label}
                value={
                  convertToPercentage(metrics.responseSeries.statusCode5xx.total, metrics.responseSeries.total) || ' - '
                }
                bgColor={'--pf-global--palette--red-100'}
                showChart={false}
              />
            </GridItem>

            {/* Chart pie distribution data traffic card and total bytes */}
            {!!metrics.responseRateSeries?.statusCode4xx.data && !!metrics.responseRateSeries?.statusCode5xx.data && (
              <>
                <GridItem span={4}>
                  <Card isFullHeight>
                    <SkChartPie
                      format={formatBytes}
                      data={[
                        {
                          x: metrics.responseSeries.statusCode4xx.label,
                          y: metrics.responseSeries.statusCode4xx.total
                        },
                        {
                          x: metrics.responseSeries.statusCode5xx.label,
                          y: metrics.responseSeries.statusCode5xx.total
                        }
                      ]}
                    />
                  </Card>
                </GridItem>

                <GridItem span={8}>
                  <Card isFullHeight>
                    <CardTitle>{MetricsLabels.ErrorRateSeriesAxisYLabel}</CardTitle>
                    <CardBody>
                      <SkChartArea
                        formatY={(y: number) => formatToDecimalPlacesIfCents(y, 3)}
                        themeColor={ChartThemeColor.orange}
                        legendLabels={[
                          metrics.responseRateSeries.statusCode4xx.label,
                          metrics.responseRateSeries.statusCode5xx.label
                        ]}
                        data={[
                          metrics.responseRateSeries.statusCode4xx.data,
                          metrics.responseRateSeries.statusCode5xx.data
                        ]}
                      />
                    </CardBody>
                  </Card>
                </GridItem>
              </>
            )}
          </>
        )}
      </>
    )}
  </>
));

export default Charts;
