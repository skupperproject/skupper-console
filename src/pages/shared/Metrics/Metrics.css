.metric-disabled {
  --pf-c-card--BackgroundColor: var(--pf-v5-global--palette--black-200) !important;
  color: var(--pf-v5-global--palette--black-500);
}
