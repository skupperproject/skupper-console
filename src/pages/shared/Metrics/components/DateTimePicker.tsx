import React, { FC, Fragment, useEffect, useState } from 'react';

import {
  CalendarMonth,
  InputGroup,
  InputGroupItem,
  TextInput,
  Button,
  Popover,
  Dropdown,
  DropdownItem,
  DropdownList,
  MenuToggle,
  MenuToggleElement
} from '@patternfly/react-core';
import { OutlinedCalendarAltIcon, OutlinedClockIcon } from '@patternfly/react-icons';
import { format } from 'date-fns';

import { formatDate, formatTime } from '../Metrics.constants';
import { DateTimePickerProps } from '../Metrics.interfaces';

const DateTimePicker: FC<DateTimePickerProps> = function ({
  defaultDate = formatDate,
  defaultTime = formatTime,
  isDisabled = false,
  startDate,
  onSelect
}) {
  const [isCalendarOpen, setIsCalendarOpen] = useState(false);
  const [isTimeOpen, setIsTimeOpen] = useState(false);
  const [valueDate, setValueDate] = useState(defaultDate);
  const [valueTime, setValueTime] = useState(defaultTime);

  const disablePreEndDates = (date: Date) => date <= new Date();
  const disablePreStartDates = (date: Date) => (startDate ? date >= startDate : true);

  const handleToggleCalendar = () => {
    setIsCalendarOpen(!isCalendarOpen);
    setIsTimeOpen(false);
  };

  const handleToggleTime = () => {
    setIsTimeOpen(!isTimeOpen);
    setIsCalendarOpen(false);
  };

  const handleSelectCalendar = (_event: React.MouseEvent<HTMLButtonElement, MouseEvent>, newValueDate: Date) => {
    const date = format(newValueDate, formatDate);
    const time = valueTime === formatTime ? '1:00' : valueTime;
    setIsCalendarOpen(!isCalendarOpen);
    setValueDate(date);
    setValueTime(time);

    if (onSelect) {
      onSelect({
        seconds: new Date(Date.parse(`${date} ${time}`)).getTime() / 1000
      });
    }
  };

  const handleSelectTime = (ev: React.MouseEvent<Element, MouseEvent> | undefined) => {
    const date = valueDate === formatDate ? format(new Date(), formatDate) : valueDate;
    const time = ev?.currentTarget?.textContent as string;

    setIsTimeOpen(!isTimeOpen);
    setValueDate(date);
    setValueTime(time);

    if (onSelect) {
      onSelect({
        seconds: new Date(Date.parse(`${date} ${time}`)).getTime() / 1000
      });
    }
  };

  useEffect(() => {
    setValueDate(defaultDate);
    setValueTime(defaultTime);
  }, [defaultDate, defaultTime]);

  const Calendar = (
    <CalendarMonth
      data-testid="date-time-picker-calendar"
      date={new Date(valueDate)}
      onChange={handleSelectCalendar}
      validators={[disablePreEndDates, disablePreStartDates]}
    />
  );

  const TimeDropdown = (
    <Dropdown
      onSelect={handleSelectTime}
      isOpen={isTimeOpen}
      isScrollable
      onOpenChange={(isOpen: boolean) => setIsTimeOpen(isOpen)}
      toggle={(toggleRef: React.Ref<MenuToggleElement>) => (
        <MenuToggle
          ref={toggleRef}
          onClick={handleToggleTime}
          isExpanded={isTimeOpen}
          icon={<OutlinedClockIcon />}
          isDisabled={isDisabled}
        />
      )}
    >
      <DropdownList>
        {Array.from(new Array(24), (_, i) => i).map((time) => (
          <Fragment key={time}>
            <DropdownItem>{`${time + 1}:00`}</DropdownItem>
            <DropdownItem>{`${time + 1}:30`}</DropdownItem>
          </Fragment>
        ))}
      </DropdownList>
    </Dropdown>
  );

  const CalendarButton = (
    <Button
      data-testid="date-time-picker-calendar-button"
      variant="control"
      onClick={handleToggleCalendar}
      isDisabled={isDisabled}
    >
      <OutlinedCalendarAltIcon />
    </Button>
  );

  return (
    <Popover
      position="bottom"
      bodyContent={Calendar}
      showClose={false}
      isVisible={isCalendarOpen}
      hasNoPadding
      hasAutoWidth
    >
      <InputGroup>
        <InputGroupItem>
          <TextInput
            type="text"
            id={`${valueDate}  ${valueTime}`}
            value={`${valueDate}  ${valueTime}`}
            readOnlyVariant="default"
            isDisabled={isDisabled}
          />
        </InputGroupItem>
        <InputGroupItem>{CalendarButton}</InputGroupItem>
        <InputGroupItem>{TimeDropdown}</InputGroupItem>
      </InputGroup>
    </Popover>
  );
};

export default DateTimePicker;
