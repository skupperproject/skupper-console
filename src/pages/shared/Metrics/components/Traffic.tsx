import { FC, useCallback, useEffect, useState } from 'react';

import { Card, CardBody, CardExpandableContent, CardHeader, CardTitle } from '@patternfly/react-core';
import { SearchIcon } from '@patternfly/react-icons';
import { useQuery } from '@tanstack/react-query';

import { isPrometheusActive } from '@config/config';
import EmptyData from '@core/components/EmptyData';

import TrafficCharts from './TrafficCharts';
import { MetricsLabels } from '../Metrics.enum';
import { QueriesMetrics, SelectedMetricFilters } from '../Metrics.interfaces';
import MetricsController from '../services';

interface TrafficProps {
  selectedFilters: SelectedMetricFilters;
  openSections?: boolean;
  forceUpdate?: number;
  refetchInterval?: number;
}

const Traffic: FC<TrafficProps> = function ({ selectedFilters, forceUpdate, refetchInterval, openSections = true }) {
  const [isExpanded, setIsExpanded] = useState(openSections);

  const { data, refetch } = useQuery(
    [QueriesMetrics.GetTraffic, selectedFilters],
    () => MetricsController.getTraffic(selectedFilters),
    {
      enabled: isPrometheusActive,
      refetchInterval,
      keepPreviousData: true
    }
  );

  const handleExpand = useCallback(() => {
    setIsExpanded(!isExpanded);
  }, [isExpanded]);

  const handleRefetchMetrics = useCallback(() => {
    isPrometheusActive && refetch();
  }, [refetch]);

  useEffect(() => {
    if (forceUpdate) {
      handleRefetchMetrics();
    }
  }, [forceUpdate, handleRefetchMetrics]);

  return (
    <Card isExpanded={isExpanded}>
      <CardHeader onExpand={handleExpand}>
        <CardTitle>{MetricsLabels.DataTransferTitle}</CardTitle>
      </CardHeader>

      <CardExpandableContent>
        <CardBody>
          {data?.txTimeSerie || data?.rxTimeSerie ? (
            <TrafficCharts byteRateData={data} />
          ) : (
            <EmptyData
              message={MetricsLabels.NoMetricFoundTitleMessage}
              description={MetricsLabels.NoMetricFoundDescriptionMessage}
              icon={SearchIcon}
            />
          )}
        </CardBody>
      </CardExpandableContent>
    </Card>
  );
};

export default Traffic;
