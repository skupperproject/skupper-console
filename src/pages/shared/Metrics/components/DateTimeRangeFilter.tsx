import { FC, useRef, useState } from 'react';

import {
  Button,
  DescriptionList,
  DescriptionListDescription,
  DescriptionListGroup,
  DescriptionListTerm,
  Divider,
  Flex,
  FlexItem,
  Menu,
  MenuContent,
  MenuItem,
  MenuToggle,
  Popover,
  Title
} from '@patternfly/react-core';
import { OutlinedClockIcon } from '@patternfly/react-icons';
import { format } from 'date-fns';

import { timeIntervalMap } from '@config/prometheus';

import DateTimePicker from './DateTimePicker';
import { formatDate, formatTime } from '../Metrics.constants';
import { SelectTimeIntervalProps } from '../Metrics.interfaces';

const DateTimeRangeFilter: FC<SelectTimeIntervalProps> = function ({
  startTimeLimit,
  onSelectTimeInterval,
  startSelected,
  endSelected,
  duration
}) {
  const popoverRef = useRef<HTMLInputElement>(null);

  const [isTimeIntervalOpen, setIsTimeIntervalOpen] = useState(false);
  const [startLabel, setStartLabel] = useState<string | undefined>(
    startSelected ? formatDateTime(startSelected) : findDurationLabel(duration)
  );
  const [endLabel, setEndLabel] = useState<string | undefined>(formatDateTime(endSelected));
  const [start, setStart] = useState<number | undefined>(startSelected);
  const [end, setEnd] = useState<number | undefined>(endSelected);

  const handleToggle = () => {
    setIsTimeIntervalOpen(!isTimeIntervalOpen);
  };

  const handleSetDateTimeStart = ({ seconds }: { label: string; seconds: number }) => {
    setStart(seconds);
    setStartLabel(formatDateTime(seconds));

    if (end && seconds > end) {
      setEnd(seconds + 60 * 30);
      setEndLabel(formatDateTime(seconds + 60 * 30));
    }
  };
  const handleSetDateTimeEnd = ({ seconds }: { label: string; seconds: number }) => {
    setEnd(seconds);
    setEndLabel(formatDateTime(seconds));

    if (start && seconds < start) {
      setStart(seconds);
      setStartLabel(formatDateTime(seconds));
    }
  };

  const handleSelectFromDateTimePicker = () => {
    setIsTimeIntervalOpen(false);

    if (onSelectTimeInterval) {
      onSelectTimeInterval({ start, end, duration: undefined });
    }
  };

  const handleSelectFromTimeInterval = (value: string) => {
    const { seconds, label } = timeIntervalMap[value as string];

    setStart(undefined);
    setEnd(undefined);
    setStartLabel(label);
    setEndLabel(undefined);
    setIsTimeIntervalOpen(false);

    if (onSelectTimeInterval) {
      onSelectTimeInterval({ start: undefined, end: undefined, duration: seconds });
    }
  };

  // The timeInterval dropdown doesn't has endLabel= undefined and show the text last min. last day ecc...
  const label = endLabel ? [startLabel, endLabel].join(' to ') : startLabel;
  const startDate = new Date(startTimeLimit / 1000 - 60 * 60 * 24 * 1000);

  return (
    <Popover
      position="left"
      bodyContent={
        <Flex>
          <FlexItem>
            <DescriptionList>
              <DescriptionListTerm>
                <Title headingLevel="h4">Absolute time range</Title>
              </DescriptionListTerm>
              <DescriptionListGroup>
                <DescriptionListTerm>From</DescriptionListTerm>
                <DescriptionListDescription>
                  <DateTimePicker
                    onSelect={handleSetDateTimeStart}
                    startDate={startDate}
                    defaultDate={formatDateTime(start, formatDate)}
                    defaultTime={formatDateTime(start, formatTime)}
                  />
                </DescriptionListDescription>
              </DescriptionListGroup>

              <DescriptionListGroup>
                <DescriptionListTerm>To</DescriptionListTerm>
                <DescriptionListDescription>
                  <DateTimePicker
                    onSelect={handleSetDateTimeEnd}
                    isDisabled={!start}
                    startDate={startDate}
                    defaultDate={formatDateTime(end, formatDate)}
                    defaultTime={formatDateTime(end, formatTime)}
                  />
                </DescriptionListDescription>
              </DescriptionListGroup>

              <DescriptionListGroup>
                <DescriptionListDescription>
                  <Button
                    variant="primary"
                    onClick={() => handleSelectFromDateTimePicker()}
                    isDisabled={!(start && end)}
                  >
                    Apply time range
                  </Button>
                </DescriptionListDescription>
              </DescriptionListGroup>
            </DescriptionList>
          </FlexItem>

          <Divider
            orientation={{
              default: 'vertical'
            }}
          />

          <FlexItem>
            <Menu onSelect={(_, value) => handleSelectFromTimeInterval(value as string)} isPlain={true} isScrollable>
              <MenuContent>{getTimeIntervals(startTimeLimit)}</MenuContent>
            </Menu>
          </FlexItem>
        </Flex>
      }
      showClose={false}
      isVisible={isTimeIntervalOpen}
      hasAutoWidth
      triggerRef={popoverRef}
    >
      <MenuToggle ref={popoverRef} onClick={handleToggle} icon={<OutlinedClockIcon />}>
        {label}
      </MenuToggle>
    </Popover>
  );
};

export default DateTimeRangeFilter;

function formatDateTime(seconds?: number, stringFormat = `${formatDate}  ${formatTime}`) {
  if (!seconds) {
    return undefined;
  }

  return format(new Date(seconds * 1000), stringFormat);
}

function getTimeIntervals(startTime: number) {
  return Object.values(timeIntervalMap)
    .filter(({ seconds }) => new Date().getTime() - seconds * 1000 > startTime / 1000)
    .map(({ key, label }, index) => (
      <MenuItem key={index} itemId={key}>
        {label}
      </MenuItem>
    ));
}

function findDurationLabel(duration?: number) {
  return (
    Object.values(timeIntervalMap).find(({ seconds }) => seconds === duration)?.label || timeIntervalMap.oneMinute.label
  );
}
