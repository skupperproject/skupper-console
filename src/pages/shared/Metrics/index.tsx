import { FC, useCallback, useState } from 'react';

import { Stack, StackItem } from '@patternfly/react-core';

import { AvailableProtocols } from '@API/REST.enum';

import MetricFilters from './components/Filters';
import Latency from './components/Latency';
import Request from './components/Request';
import Traffic from './components/Traffic';
import { MetricsProps, QueryMetricsParams } from './Metrics.interfaces';

const Metrics: FC<MetricsProps> = function ({
  configFilters,
  defaultMetricFilterValues,
  sourceSites,
  destSites,
  sourceProcesses,
  destProcesses,
  availableProtocols,
  startTimeLimit, // Use startTimeLimit to set the left temporal limit of the SelectTimeInterval filter
  openSections,
  onGetMetricFilters
}) {
  const { refreshDataInterval: defaultRefreshDataInterval, ...filters } = defaultMetricFilterValues;

  const [queryParams, setQueryParams] = useState<QueryMetricsParams>(filters);
  const [refetchInterval, setRefetchInterval] = useState(defaultRefreshDataInterval);
  const [shouldUpdateData, setShouldUpdateData] = useState(0);

  //Filters: refetch manually the prometheus API
  const handleShouldUpdateData = useCallback(() => {
    setShouldUpdateData(new Date().getTime());
  }, []);

  // Filters: Set the prometheus query params with the filter values
  const handleUpdateQueryParams = useCallback(
    (updatedFilters: QueryMetricsParams, refreshDataInterval?: number) => {
      setRefetchInterval(refreshDataInterval);
      setQueryParams(updatedFilters);
      if (onGetMetricFilters) {
        onGetMetricFilters({ ...updatedFilters, refreshDataInterval });
      }
    },
    [onGetMetricFilters]
  );

  return (
    <Stack hasGutter>
      <StackItem style={{ position: 'sticky', top: 0, zIndex: 1 }}>
        <MetricFilters
          configFilters={configFilters}
          defaultMetricFilterValues={defaultMetricFilterValues}
          defaultRefreshDataInterval={defaultRefreshDataInterval}
          sourceSites={sourceSites}
          destSites={destSites}
          sourceProcesses={sourceProcesses}
          destProcesses={destProcesses}
          availableProtocols={availableProtocols}
          startTimeLimit={startTimeLimit}
          isRefetching={false}
          onRefetch={handleShouldUpdateData}
          onSelectFilters={handleUpdateQueryParams}
        />
      </StackItem>

      <StackItem>
        <Traffic selectedFilters={queryParams} forceUpdate={shouldUpdateData} refetchInterval={refetchInterval} />
      </StackItem>

      {queryParams.protocol !== AvailableProtocols.Tcp && (
        <>
          <StackItem>
            <Latency
              selectedFilters={queryParams}
              openSections={openSections?.latency}
              forceUpdate={shouldUpdateData}
              refetchInterval={refetchInterval}
            />
          </StackItem>

          <StackItem>
            <Request
              selectedFilters={queryParams}
              openSections={openSections?.request}
              forceUpdate={shouldUpdateData}
              refetchInterval={refetchInterval}
            />
          </StackItem>
        </>
      )}
    </Stack>
  );
};

export default Metrics;
