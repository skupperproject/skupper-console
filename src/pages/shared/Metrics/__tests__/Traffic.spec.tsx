import { Suspense } from 'react';

import * as ReactQuery from '@tanstack/react-query';
import { render, screen, waitForElementToBeRemoved } from '@testing-library/react';
import { Server } from 'miragejs';

import { ProcessResponse } from '@API/REST.interfaces';
import { waitForElementToBeRemovedTimeout } from '@config/config';
import { getTestsIds } from '@config/testIds';
import { Wrapper } from '@core/components/Wrapper';
import processesData from '@mocks/data/PROCESSES.json';
import { loadMockServer } from '@mocks/server';
import LoadingPage from '@pages/shared/Loading';

import Traffic from '../components/Traffic';
import { MetricsLabels } from '../Metrics.enum';

let component;
const processResult = processesData.results[0] as ProcessResponse;

describe('Traffic component', () => {
  let server: Server;
  beforeEach(() => {
    server = loadMockServer() as Server;
    server.logging = false;
  });

  afterEach(() => {
    server.shutdown();
    jest.clearAllMocks();
  });

  it('should render the Traffic section of the metric', async () => {
    component = render(
      <Wrapper>
        <Suspense fallback={<LoadingPage />}>
          <Traffic
            selectedFilters={{
              sourceProcess: processResult.name
            }}
            openSections={true}
            forceUpdate={1}
          />
        </Suspense>
      </Wrapper>
    );

    await waitForElementToBeRemoved(() => screen.queryByTestId(getTestsIds.loadingView()), {
      timeout: waitForElementToBeRemovedTimeout
    });

    expect(screen.getByText(MetricsLabels.DataTransferTitle)).toBeInTheDocument();
    expect(component).toMatchSnapshot();
  });

  it('should render the Traffic section and display the no metric found message', async () => {
    jest.spyOn(ReactQuery, 'useQuery').mockImplementation(jest.fn().mockReturnValue({ data: null }));

    render(
      <Wrapper>
        <Suspense fallback={<LoadingPage />}>
          <Traffic
            selectedFilters={{
              sourceProcess: processResult.name
            }}
          />
        </Suspense>
      </Wrapper>
    );

    expect(screen.getByText(MetricsLabels.NoMetricFoundTitleMessage)).toBeInTheDocument();
    expect(screen.getByText(MetricsLabels.NoMetricFoundDescriptionMessage)).toBeInTheDocument();
  });
});
