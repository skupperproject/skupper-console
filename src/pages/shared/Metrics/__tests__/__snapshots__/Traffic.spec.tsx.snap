// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Traffic component should render the Traffic section of the metric 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="pf-v5-c-card"
        data-ouia-component-id="OUIA-Generated-Card-1"
        data-ouia-component-type="PF5/Card"
        data-ouia-safe="true"
        id=""
      >
        <div
          class="pf-v5-c-card__header"
        >
          <div
            class="pf-v5-c-card__header-toggle"
          >
            <button
              aria-disabled="false"
              class="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-1"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              type="button"
            >
              <span
                class="pf-v5-c-card__header-toggle-icon"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            class="pf-v5-c-card__header-main"
          >
            <div
              class="pf-v5-c-card__title"
            >
              <div
                class="pf-v5-c-card__title-text"
              >
                Traffic
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <svg
      height="300"
      style="position: fixed; top: -9999px; left: -9999px;"
      viewBox="0 0 300 300"
      width="300"
      xlink="http://www.w3.org/1999/xlink"
    >
      <text />
    </svg>
  </body>,
  "container": <div>
    <div
      class="pf-v5-c-card"
      data-ouia-component-id="OUIA-Generated-Card-1"
      data-ouia-component-type="PF5/Card"
      data-ouia-safe="true"
      id=""
    >
      <div
        class="pf-v5-c-card__header"
      >
        <div
          class="pf-v5-c-card__header-toggle"
        >
          <button
            aria-disabled="false"
            class="pf-v5-c-button pf-m-plain"
            data-ouia-component-id="OUIA-Generated-Button-plain-1"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
            type="button"
          >
            <span
              class="pf-v5-c-card__header-toggle-icon"
            >
              <svg
                aria-hidden="true"
                class="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 256 512"
                width="1em"
              >
                <path
                  d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="pf-v5-c-card__header-main"
        >
          <div
            class="pf-v5-c-card__title"
          >
            <div
              class="pf-v5-c-card__title-text"
            >
              Traffic
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
